lib/configuration.js

```diff
@@ -15,150 +15,430 @@
- *  set to production: if the flag is set to 'production' then the
- *  `shouldReportErrorsToAPI` property will be set to true and the API client
- *  will attempt to send errors to the Stackdriver Error Reporting API.
+ * The _gatherLocalConfiguration function is responsible for determining
+ * directly determing whether the properties `reportUncaughtExceptions` and
+ * `key`, which can be optionally supplied in the runtime configuration, should
+ * be merged into the instance. This function also calls several specialized
+ * environmental variable checkers which not only check for the optional runtime
+ * configuration supplied values but also the processes environmental values.
+ * @memberof Configuration
+ * @private
+ * @function _gatherLocalConfiguration
+ * @returns {Undefined} - does not return anything
  */
-
+Configuration.prototype._gatherLocalConfiguration = function() {
+  if (isPlainObject(this._givenConfiguration)) {
+    if (isBoolean(this._givenConfiguration.reportUncaughtExceptions)) {
+      this._reportUncaughtExceptions = this._givenConfiguration
+        .reportUncaughtExceptions;
+    }
+    if (isString(this._givenConfiguration.key)) {
+      this._key = this._givenConfiguration.key;
+    }
+  }
+};
 /**
- * Orchestrates the gathering of configuration based on submission of the
- * initConfiguration parameter. If all optional properties are given in the
- * initConfiguration parameter then the only environmental variable that will
- * be read is the NODE_ENV variable. If initConfiguration is not supplied or
- * only partially supplied then gatherConfiguration will attempt to read
- * environmental variables to supplement information as necessary.
- * @function gatherConfiguration
- * @param {ConfigurationOptions} initConfiguration - the options object
- * @returns {NormalizedConfigurationVariables} - a configuration object
- *  supplemented with environmental information as necessary
+ * The callback for `utils.getProjectNumber` this function is responsible for
+ * determining whether or not an error occured in the metadata transaction and,
+ * if one did not, assigning the value of the transaction (the project number)
+ * to the instance property `_projectNumber`.
+ * @memberof Configuration
+ * @private
+ * @function _assimilateProjectNumber
+ * @param {Error|Undefined} err - If applicable, the error that occured during 
+ *  the transaction
+ * @param {Null|String} projectNum - The project number if successfully 
+ *  retrieved, null otherwise.
+ * @returns {Undefined} - does not return anything
  */
-function gatherConfiguration(initConfiguration) {
-
-  var projectId = null;
-  var reportUncaughtExceptions = true;
-  var serviceContext = null;
-  var key = null;
-  var shouldReportErrorsToAPI = false;
-
-  if (env.NODE_ENV === 'production') {
-
-    shouldReportErrorsToAPI = true;
+Configuration.prototype._assimilateProjectNumber = function(err, projectNum) {
+  if (isNull(err) && isString(projectNum)) {
```

Comment:
My general understanding of the contract of an "errorback" style function is that if the error value is truthy then an error occurred and the other parameters should not be considered. Otherwise, the error value will be falsey to indicate that no error occured. The specific falsey or truthy values passed to the first argument are implementation details and a typecheck like this will start failing without violating the errorback contract if a different falsey value is specified. Since `''` is a falsey value, I would consider this as no error occurring and a bug on the calling end if it should be treated as such.